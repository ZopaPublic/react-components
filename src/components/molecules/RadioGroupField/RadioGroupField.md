### Summary

`<RadioFieldGroup />` is a convenient wrapper that renders group of `<RadioField>`

⚠️ &nbsp;Note that the `value` for each item has to be unique as it determines the `id`
and `htmlFor` prop fof each radio field.

### Examples

- Uncontrolled

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  items={[
    { value: 'one', label: 'label one' },
    { value: 'two', label: 'label two' },
  ]}
  onChange={(value) => console.log(value)}
  label="Uncontrolled"
/>;
```

- Uncontrolled with one radio checked by default

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  items={[
    { value: 'three', label: 'label three' },
    { value: 'four', label: 'label four', defaultChecked: true },
  ]}
  onChange={(value) => console.log(value)}
  label="Uncontrolled with one radio checked by default"
/>;
```

- Uncontrolled with hidden control

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  items={[
    { value: 'three', label: 'label three' },
    { value: 'four', label: 'label four', defaultChecked: true },
  ]}
  onChange={(value) => console.log(value)}
  label="Uncontrolled with hidden control"
  hideControl
/>;
```

- Controlled

```js
import { useState } from 'react';
import { RadioGroupField } from '@zopauk/react-components';

const ControlledRadioGroupField = () => {
  const [value, setValue] = useState('five');
  return (
    <RadioGroupField
      items={[
        { value: 'five', label: 'label five' },
        { value: 'six', label: 'label six' },
      ]}
      value={value}
      onChange={setValue}
      label="Controlled"
    />
  );
};

<ControlledRadioGroupField />;
```

- Disabled

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  items={[
    { value: 'seven', label: 'label seven' },
    { value: 'eight', label: 'label eight' },
  ]}
  onChange={(value) => console.log(value)}
  label="Disabled"
  disabled
/>;
```

- Disabled and valid

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  items={[
    { value: 'eight', label: 'label eight' },
    { value: 'nine', label: 'label nine', defaultChecked: true },
  ]}
  onChange={(value) => console.log(value)}
  label="Disabled and valid"
  isValid={true}
  disabled
/>;
```

- With error

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  items={[
    { value: 'ten', label: 'label ten' },
    { value: 'eleven', label: 'label eleven' },
  ]}
  onChange={(value) => console.log(value)}
  label="With error"
  errorMessage="You have to pick one"
/>;
```

- Responsive

```jsx
import { RadioGroupField } from '@zopauk/react-components';

<RadioGroupField
  flexColProps={{
    xs: 12,
    m: 6,
    xl: 3,
  }}
  items={[
    { value: 'fourteen', label: 'label fourteen' },
    { value: 'fifteen', label: 'label fifteen' },
    { value: 'sixteen', label: 'label sixteen' },
    { value: 'seventeen', label: 'label seventeen' },
  ]}
  onChange={(value) => console.log(value)}
  label="Responsive"
  name="Responsive"
  data-automation="testing"
/>;
```
