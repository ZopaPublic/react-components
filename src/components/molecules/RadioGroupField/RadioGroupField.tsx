import React, { useState } from 'react';
import styled from 'styled-components';
import FlexRow, { FlexRowProps } from '../../layout/FlexRow/FlexRow';
import FlexCol, { FlexColProps } from '../../layout/FlexCol/FlexCol';
import RadioField from '../RadioField/RadioField';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import Fieldset from '../../atoms/Fieldset/Fieldset';
import Legend from '../../atoms/Legend/Legend';

const RadioWrapper = styled.div`
  padding: 4px 0;
`;

type RadioGroupFieldItem = {
  value: string;
  label?: string | React.ReactNode;
  defaultChecked?: boolean;
  disabled?: boolean;
};

export type RadioGroupFieldProps = {
  label?: string;
  items: RadioGroupFieldItem[];
  onChange: (value: string) => void;
  onBlur?: () => void;
  disabled?: boolean;
  value?: string;
  isValid?: boolean;
  errorMessage?: string;
  className?: string;
  flexColProps?: FlexColProps;
  flexRowProps?: FlexRowProps;
  hideControl?: boolean;
  name?: string;
  'data-automation'?: string;
};

const RadioGroupField = ({
  items,
  label,
  name,
  onChange,
  onBlur,
  value,
  disabled,
  isValid = false,
  errorMessage,
  className,
  flexRowProps = {},
  flexColProps = {},
  hideControl,
  'data-automation': dataAutomation,
}: RadioGroupFieldProps) => {
  const { value: defaultValue } = items.find(({ defaultChecked }) => defaultChecked) || {};

  const [innerValue, setInnerValue] = useState(defaultValue);

  const isControlled = !!value;

  const handleChange = (value: string) => () => {
    if (!disabled) {
      if (!isControlled) {
        setInnerValue(value);
      }
      onChange(value);
    }
  };

  return (
    <Fieldset className={className}>
      {label && <Legend>{label}</Legend>}
      <FlexRow {...flexRowProps}>
        {items.map((item, index) => {
          const checked = isControlled ? value === item.value : innerValue === item.value;
          return (
            <FlexCol {...flexColProps} key={item.value}>
              <RadioWrapper>
                <RadioField
                  disabled={disabled}
                  value={item.value}
                  onChange={handleChange(item.value)}
                  onBlur={onBlur}
                  label={item.label}
                  checked={checked}
                  isValid={checked && isValid}
                  hideControl={hideControl}
                  name={name}
                  data-automation={dataAutomation ? `${dataAutomation}-${index}` : undefined}
                  groupLabel={label}
                />
              </RadioWrapper>
            </FlexCol>
          );
        })}
      </FlexRow>
      {errorMessage && <ErrorMessage className="mt-1">{errorMessage}</ErrorMessage>}
    </Fieldset>
  );
};

export default RadioGroupField;
