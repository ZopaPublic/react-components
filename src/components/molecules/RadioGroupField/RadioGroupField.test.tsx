import React, { useState } from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import axe from '../../../../axe-helper';
import RadioGroupField from './RadioGroupField';

describe('<RadioGroupField />', () => {
  it('should handle value change if it is not controlled', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <RadioGroupField
        label="label"
        onChange={onChange}
        items={[
          {
            value: 'one',
            label: 'label one',
            defaultChecked: true,
          },
          {
            value: 'two',
            label: 'label two',
          },
        ]}
      />,
    );
    const labelOne: any = getByLabelText('label one');
    const labelTwo: any = getByLabelText('label two');
    expect(labelOne.checked).toEqual(true);
    expect(labelTwo.checked).toEqual(false);

    fireEvent.click(labelTwo);
    expect(onChange).toHaveBeenCalledWith('two');
    expect(labelOne.checked).toEqual(false);
    expect(labelTwo.checked).toEqual(true);
  });

  it('should not change value if it is disabled', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <RadioGroupField
        disabled
        label="label"
        onChange={onChange}
        items={[
          {
            value: 'one',
            label: 'label one',
          },
          {
            value: 'two',
            label: 'label two',
          },
        ]}
      />,
    );
    const labelOne = getByLabelText('label one');
    const labelTwo: any = getByLabelText('label two');

    if (labelOne instanceof HTMLInputElement && labelTwo instanceof HTMLInputElement) {
      expect(labelOne.checked).toEqual(false);
      expect(labelTwo.checked).toEqual(false);
      fireEvent.click(labelOne);
      expect(onChange).not.toHaveBeenCalledWith('one');
      expect(labelOne.checked).toEqual(false);
      expect(labelOne.checked).toEqual(false);

      fireEvent.click(labelTwo);
      expect(onChange).not.toHaveBeenCalledWith('two');
      expect(labelOne.checked).toEqual(false);
      expect(labelTwo.checked).toEqual(false);
    } else {
      throw new Error('expected labelOne and labelTwo to be html input elements');
    }
  });

  it('should handle value change if it is controlled', () => {
    const ControlledRadioGroupField = () => {
      const [value, setValue] = useState('one');
      return (
        <RadioGroupField
          items={[
            { value: 'one', label: 'label one' },
            { value: 'two', label: 'label two' },
          ]}
          value={value}
          onChange={setValue}
          label="label"
        />
      );
    };
    const { getByLabelText } = render(<ControlledRadioGroupField />);
    const labelOne: any = getByLabelText('label one');
    const labelTwo: any = getByLabelText('label two');
    expect(labelOne.checked).toEqual(true);
    expect(labelTwo.checked).toEqual(false);
    act(() => {
      fireEvent.click(labelTwo);
    });
    expect(labelOne.checked).toEqual(false);
    expect(labelTwo.checked).toEqual(true);
  });

  it('renders the component with props with no a11y violations', async () => {
    const { container } = render(
      <RadioGroupField
        label="label"
        onChange={jest.fn()}
        items={[
          { value: 'one', label: 'label one' },
          { value: 'two', label: 'label two' },
        ]}
        data-automation="ZA.test"
      />,
    );
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('renders the component with no control icon and no a11y violations', async () => {
    const { container } = render(
      <RadioGroupField
        label="label"
        onChange={jest.fn()}
        items={[
          { value: 'one', label: 'label one' },
          { value: 'two', label: 'label two' },
        ]}
        hideControl
      />,
    );
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('should have change of one group when multiple groups with same values', () => {
    const ControlledRadioGroupField = () => {
      const [value1, setValue1] = useState('one');
      const [value2, setValue2] = useState('one');

      return (
        <>
          <RadioGroupField
            label="Group 1"
            value={value1}
            onChange={setValue1}
            items={[
              { value: 'one', label: 'label one' },
              { value: 'two', label: 'label two' },
            ]}
            hideControl
          />
          <RadioGroupField
            label="Group 2"
            value={value2}
            onChange={setValue2}
            items={[
              { value: 'one', label: 'label one' },
              { value: 'two', label: 'label two' },
            ]}
            hideControl
          />
        </>
      );
    };
    const { container } = render(<ControlledRadioGroupField />);
    const inputOne: any = container.querySelector('#radio-id-Group-2-one');
    const inputTwo: any = container.querySelector('#radio-id-Group-2-two');

    const inputThree: any = container.querySelector('#radio-id-Group-1-one');
    const inputFour: any = container.querySelector('#radio-id-Group-1-two');

    expect(inputOne.checked).toEqual(true);
    expect(inputTwo.checked).toEqual(false);
    expect(inputThree.checked).toEqual(true);
    expect(inputFour.checked).toEqual(false);
    act(() => {
      fireEvent.click(inputTwo);
    });
    expect(inputOne.checked).toEqual(false);
    expect(inputTwo.checked).toEqual(true);
    expect(inputThree.checked).toEqual(true);
    expect(inputFour.checked).toEqual(false);

    act(() => {
      fireEvent.click(inputFour);
    });

    expect(inputOne.checked).toEqual(false);
    expect(inputTwo.checked).toEqual(true);
    expect(inputThree.checked).toEqual(false);
    expect(inputFour.checked).toEqual(true);
  });
});
