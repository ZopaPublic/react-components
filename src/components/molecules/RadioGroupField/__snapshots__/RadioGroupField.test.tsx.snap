// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroupField /> renders the component with no control icon and no a11y violations 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

.c8 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c5 {
  max-width: 100%;
  width: 100%;
}

.c6 {
  position: relative;
  margin-bottom: 10px;
}

.c6:last-of-type {
  margin-bottom: 0;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.4;
  font-weight: 700;
  font-size: 16px;
  color: #2C3236;
  padding: 14px 16px;
  border: 1px solid #818F9B;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  border-radius: 8px;
  position: relative;
  margin-bottom: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c9:before {
  content: '';
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  border-radius: 50%;
  height: 20px;
  width: 20px;
  display: inline-block;
  margin-right: 8px;
  border: 1px #818F9B solid;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c9:after {
  content: '';
  position: absolute;
  border-radius: 50%;
  display: inline-block;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  left: 21px;
  top: 19px;
}

.c9:hover {
  cursor: pointer;
}

.c7 {
  width: 1px;
  height: 1px;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c7:hover:not(:disabled) + label,
.c7:focus + label {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c7:hover:not(:disabled) + label:before,
.c7:focus + label:before {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c7:hover:checked:not(:disabled) + label:after,
.c7:focus:checked + label:after {
  background-color: #009E8E;
}

.c7:disabled + label {
  cursor: not-allowed;
  color: #818F9B;
  background-color: #F7F7F7;
}

.c7:disabled:not(:checked) + label {
  border-color: #D4D7D9;
}

.c7:disabled:not(:checked) + label:before {
  border-color: #D4D7D9;
}

.c7:checked + label {
  border-color: #009E8E;
  background-color: #ECFCF9;
  box-shadow: none;
}

.c7:checked + label:before {
  border-color: #009E8E;
  box-shadow: none;
}

.c7:checked + label:after {
  background-color: #009E8E;
  height: 10px;
  width: 10px;
  -webkit-animation: ciAMHu 200ms ease-in-out;
  animation: ciAMHu 200ms ease-in-out;
}

.c7 + label:before,
.c7 + label:after {
  display: none;
}

.c0 {
  border: 0;
  padding: 0.01em 0 0 0;
  margin: 0;
  min-width: 0;
}

.c1 {
  padding: 0;
  display: table;
  margin-bottom: 8px;
  font-weight: 600;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c4 {
  padding: 4px 0;
}

<fieldset
  class="c0"
>
  <legend
    class="c1"
  >
    label
  </legend>
  <div
    class="c2"
    cols="12"
    direction="row"
  >
    <div
      class="c3"
      cols="12"
    >
      <div
        class="c4"
      >
        <div
          class="c5 c6"
        >
          <input
            class="c7"
            id="radio-id-label-one"
            type="radio"
            value="one"
          />
          <label
            class="c8 c9"
            for="radio-id-label-one"
          >
            label one
          </label>
        </div>
      </div>
    </div>
    <div
      class="c3"
      cols="12"
    >
      <div
        class="c4"
      >
        <div
          class="c5 c6"
        >
          <input
            class="c7"
            id="radio-id-label-two"
            type="radio"
            value="two"
          />
          <label
            class="c8 c9"
            for="radio-id-label-two"
          >
            label two
          </label>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`<RadioGroupField /> renders the component with props with no a11y violations 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

.c8 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c5 {
  max-width: 100%;
  width: 100%;
}

.c6 {
  position: relative;
  margin-bottom: 10px;
}

.c6:last-of-type {
  margin-bottom: 0;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.4;
  font-weight: 400;
  font-size: 16px;
  color: #2C3236;
  padding: 14px 16px;
  border: 1px solid #818F9B;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  border-radius: 8px;
  position: relative;
  margin-bottom: 0;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
}

.c9:before {
  content: '';
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  border-radius: 50%;
  height: 20px;
  width: 20px;
  display: inline-block;
  margin-right: 8px;
  border: 1px #818F9B solid;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c9:after {
  content: '';
  position: absolute;
  border-radius: 50%;
  display: inline-block;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  left: 21px;
  top: 19px;
}

.c9:hover {
  cursor: pointer;
}

.c7 {
  width: 1px;
  height: 1px;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c7:hover:not(:disabled) + label,
.c7:focus + label {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c7:hover:not(:disabled) + label:before,
.c7:focus + label:before {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c7:hover:checked:not(:disabled) + label:after,
.c7:focus:checked + label:after {
  background-color: #009E8E;
}

.c7:disabled + label {
  cursor: not-allowed;
  color: #818F9B;
  background-color: #F7F7F7;
}

.c7:disabled:not(:checked) + label {
  border-color: #D4D7D9;
}

.c7:disabled:not(:checked) + label:before {
  border-color: #D4D7D9;
}

.c7:checked + label {
  border-color: #009E8E;
  background-color: #ECFCF9;
  box-shadow: none;
}

.c7:checked + label:before {
  border-color: #009E8E;
  box-shadow: none;
}

.c7:checked + label:after {
  background-color: #009E8E;
  height: 10px;
  width: 10px;
  -webkit-animation: ciAMHu 200ms ease-in-out;
  animation: ciAMHu 200ms ease-in-out;
}

.c0 {
  border: 0;
  padding: 0.01em 0 0 0;
  margin: 0;
  min-width: 0;
}

.c1 {
  padding: 0;
  display: table;
  margin-bottom: 8px;
  font-weight: 600;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c4 {
  padding: 4px 0;
}

<fieldset
  class="c0"
>
  <legend
    class="c1"
  >
    label
  </legend>
  <div
    class="c2"
    cols="12"
    direction="row"
  >
    <div
      class="c3"
      cols="12"
    >
      <div
        class="c4"
      >
        <div
          class="c5 c6"
        >
          <input
            class="c7"
            data-automation="ZA.test-0"
            id="radio-id-label-one"
            type="radio"
            value="one"
          />
          <label
            class="c8 c9"
            for="radio-id-label-one"
          >
            label one
          </label>
        </div>
      </div>
    </div>
    <div
      class="c3"
      cols="12"
    >
      <div
        class="c4"
      >
        <div
          class="c5 c6"
        >
          <input
            class="c7"
            data-automation="ZA.test-1"
            id="radio-id-label-two"
            type="radio"
            value="two"
          />
          <label
            class="c8 c9"
            for="radio-id-label-two"
          >
            label two
          </label>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
